File: backend/app/Enums/RideStatus.php
<?php

namespace App\Enums;

enum RideStatus: string
{
    case Pending = 'pending';
    case DriverAssigned = 'driver_assigned';
    case DriverArriving = 'driver_arriving';
    case DriverArrived = 'driver_arrived';
    case PassengerOnBoard = 'passenger_on_board';
    case InProgress = 'in_progress';
    case Completed = 'completed';
    case CancelledByPassenger = 'cancelled_by_passenger';
    case CancelledByDriver = 'cancelled_by_driver';
    case Expired = 'expired';

    public function isTerminal(): bool
    {
        return in_array($this, [
            self::Completed,
            self::CancelledByPassenger,
            self::CancelledByDriver,
            self::Expired,
        ], true);
    }

    public function label(): string
    {
        return match ($this) {
            self::Pending => 'Pending',
            self::DriverAssigned => 'Driver Assigned',
            self::DriverArriving => 'Driver Arriving',
            self::DriverArrived => 'Driver Arrived',
            self::PassengerOnBoard => 'Passenger On Board',
            self::InProgress => 'In Progress',
            self::Completed => 'Completed',
            self::CancelledByPassenger => 'Cancelled by Passenger',
            self::CancelledByDriver => 'Cancelled by Driver',
            self::Expired => 'Expired',
        };
    }
}

File: packages/taxigo_core/lib/domain/enums/ride_status.dart
/// Ride lifecycle statuses aligned with the backend API.
enum RideStatus {
  pending('pending'),
  driverAssigned('driver_assigned'),
  driverArriving('driver_arriving'),
  driverArrived('driver_arrived'),
  passengerOnBoard('passenger_on_board'),
  inProgress('in_progress'),
  completed('completed'),
  cancelledByPassenger('cancelled_by_passenger'),
  cancelledByDriver('cancelled_by_driver'),
  expired('expired');

  const RideStatus(this.value);

  final String value;

  static RideStatus fromString(String? raw) {
    return RideStatus.values.firstWhere(
      (status) => status.value == raw,
      orElse: () => RideStatus.pending,
    );
  }

  /// Localization key in [AppLocalizations] for this status.
  String get displayKey => switch (this) {
        RideStatus.pending => 'rideStatusPending',
        RideStatus.driverAssigned => 'rideStatusDriverAssigned',
        RideStatus.driverArriving => 'rideStatusDriverArriving',
        RideStatus.driverArrived => 'rideStatusDriverArrived',
        RideStatus.passengerOnBoard => 'rideStatusPassengerOnBoard',
        RideStatus.inProgress => 'rideStatusInProgress',
        RideStatus.completed => 'rideStatusCompleted',
        RideStatus.cancelledByPassenger => 'rideStatusCancelledByPassenger',
        RideStatus.cancelledByDriver => 'rideStatusCancelledByDriver',
        RideStatus.expired => 'rideStatusExpired',
      };

  bool get isTerminal => switch (this) {
        RideStatus.completed ||
        RideStatus.cancelledByPassenger ||
        RideStatus.cancelledByDriver ||
        RideStatus.expired =>
          true,
        _ => false,
      };

  bool get isActive => !isTerminal;
}
